matching/src/main.ts

```diff
@@ -0,0 +1,308 @@
+    title: data.title,
+    DOI: data.DOI,
+    journal: data["container-title"],
+    abbreviated_journal: data["short-container-title"],
+    volume: data.volume,
+    type: data.type,
+    year: getYear(data),
+  };
+
+  if (data.author?.length > 0) {
+    const { authors, firstAuthorName } = data.author.reduce(
+      (acc, author) => {
+        if (author.sequence === "first") {
+          acc.firstAuthorName = author.family;
+        }
+        acc.authors.push(author.family);
+        return acc;
+      },
+      { authors: [] as string[], firstAuthorName: "" }
+    );
+
+    obj.first_author = firstAuthorName;
+    obj.author = authors.join(" ");
+  }
+
+  // parse page metadata to get the first page only
+  if (data.page) {
+    const firstPage = data.page.split(/,|-| /g)[0];
+    if (firstPage) {
+      obj.first_page = firstPage;
+    }
+  }
+
+  if (data.issue) {
+    obj.issue = data.issue;
+  }
+
+  // bibliograohic field is the concatenation of usual bibliographic metadata
+  const biblio = buildBibliographicField(obj);
+  if (biblio.length > 0) {
+    obj.bibliographic = biblio;
+  }
+
+  return obj as BiblObj;
+};
+
+/**
+ * Extracts the year from the different alternatives.
+ * Goes in the order "issued", "published-online", "published-print" and lastly "created"
+ * @param data The data to extract the year from
+ */
+const getYear = (data: RawCrossrefData): number => {
+  // year is a date part (first one) in issued or created or published-online (we follow this order)
+  if (data.issued?.["date-parts"]) {
+    return data.issued["date-parts"][0][0];
+  } else if (data["published-online"]?.["date-parts"]) {
+    return data["published-online"]["date-parts"][0][0];
+  } else if (data["published-print"]?.["date-parts"]) {
+    return data["published-print"]["date-parts"][0][0];
+  } else if (data.created?.["date-parts"]) {
+    // this is deposit date, normally we will never use it, but it will ensure
+    // that we always have a date as conservative fallback
+    data.created["date-parts"][0][0];
+  } else {
+    return null;
+  }
+};
+
+const index = async (options: Options, client: Client) => {
+  const readStream = fs
+    .createReadStream(options.dump)
+    .pipe(createDecompressor())
+    .pipe(split())
+    .pipe(
+      map((rawData: any, cb: any) => {
+        cb(null, createBiblObj(massage(rawData)));
+      })
+    )
+    .on("error", (error) => console.error("Error occurred: " + error))
+    .on("finish", () => {
+      console.log("Finished. ");
+      const execTime = Date.now() - options.start.getTime();
+      console.info("Execution time: %dms", execTime);
+    });
+
+  let i = 0;
+  let indexed = 0;
+  let batch: (ElasticIndexHeader | BiblObj)[] = [];
+
+  readStream.on("data", async (doc: BiblObj) => {
+    // filter some type of DOI not corresponding to a publication (e.g. component
+    // of a publication)
+    if (filterType(doc)) {
+      return;
+    }
+
+    delete doc._id;
```

Comment:
to able to synchronize incrementally with crossref, I think doi could be used as identifier, @kermitt2 @karatekaneen WDYT ?

Reply:
It sounds like a very simple solution that may help towards the incremental synchronization that I really long for. But I have a _very_ limited view on how all this works in the rest of glutton.

Reply:
so far biblio glutton has no tool to do synchronization, but I'm trying to work on something and I need to confirm the usage of doi as unique identifier (I have tried to rebuild the index using doi, and it seems there is no data loss..)

Reply:
Interesting, I've been meaning to look in to that myself. 
Do you mean that the internal LMDB's does not clear out when you try to add a batch of fresh data? And if so, using doi here would make a lot of sense.